src/WrappedMToken.sol

```diff
@@ -580,18 +577,13 @@ contract WrappedMToken is IWrappedMToken, Migratable, ERC20Extended {
      * @return unwrapped_ The amount of M withdrawn.
      */
     function _unwrap(address account_, address recipient_, uint240 amount_) internal returns (uint240 unwrapped_) {
-        _burn(account_, amount_);
-
-        uint240 startingBalance_ = _mBalanceOf(address(this));
-
-        // NOTE: The behavior of `IMTokenLike.transfer` is known, so its return can be ignored.
-        IMTokenLike(mToken).transfer(recipient_, _getSafeTransferableM(amount_, currentIndex()));
-
         // NOTE: When this WrappedMToken contract is earning, any amount of M sent from it is converted to a principal
         //       amount at the MToken contract, which when represented as a present amount, may be a rounding error
-        //       amount more than `amount_`. In order to capture the real decrease in M, the difference between the
-        //       ending and starting M balance is returned.
-        return startingBalance_ - _mBalanceOf(address(this));
+        //       amount more than `amount_`. The real decrease in M may be larger than `amount_`.
+        _burn(account_, unwrapped_ = amount_);
+
+        // NOTE: The behavior of `IMTokenLike.transfer` is known, so its return can be ignored.
+        IMTokenLike(mToken).transfer(recipient_, _getSufficientTransferableM(recipient_, amount_));
```

Comment:
necessity of `_getSufficientTransferableM` is not clear, round up will handled by the M itself https://github.com/m0-foundation/protocol/blob/main/src/MToken.sol#L385

and if receiver is an earner, they will receive rounded up amount as well (not precise, but in favor of user)